### PR TITLE
Add asserts & test for #20958

### LINKED ISSUE #20958: Assertion failed 'structType != NO_CLASS_HANDLE' in 'System.SpanTests.IndexerTests:IndexerWithRangeTest()
https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/job/x64_checked_ubuntu_corefx_baseline/751/
https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/job/arm64_cross_checked_ubuntu16.04_corefx_baseline_tst/37/
https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/job/arm_cross_checked_ubuntu_corefx_baseline_tst/28/

```
Assert failure(PID 2702 [0x00000a8e], Thread: 3226 [0x0c9a]): Assertion failed 'structType != NO_CLASS_HANDLE' in 'System.SpanTests.IndexerTests:IndexerWithRangeTest()' (IL size 204)

    File: /mnt/j/workspace/dotnet_coreclr/master/jitstress/arm_cross_checked_ubuntu_corefx_baseline/src/jit/importer.cpp Line: 768
    Image: /ssd/j/workspace/dotnet_coreclr/master/jitstress/arm_cross_checked_ubuntu_corefx_baseline_tst/_/fx/artifacts/bin/testhost/netcoreapp-Linux-Release-arm/dotnet
```

/cc @dotnet/jit-contrib 

The test assembly was excluded from Linux/arm and Linux/arm64 jobs in https://github.com/dotnet/coreclr/pull/20962
When this is fixed please update **tests/arm/corefx_linux_test_exclusions.txt** and **tests/arm64/corefx_linux_test_exclusions.txt** files

### PR DESCRIPTION
**Pull request by CarolEidt**
Add asserts for the types of the Span indexers.

Also, add a test for #20958

### CODE REVIEW

File: src/jit/importer.cpp [hunk 1]
```diff
@@ -3805,6 +3805,14 @@ GenTree* Compiler::impIntrinsic(GenTree*                newobjThis,
             JITDUMP("\nimpIntrinsic: Expanding %sSpan<T>.get_Item, T=%s, sizeof(T)=%u\n", isReadOnly ? "ReadOnly" : "",
                     info.compCompHnd->getClassName(spanElemHnd), elemSize);
 
+            if (varTypeIsStruct(impStackTop(0).val))
```
Comment (jkotas) [inline src/jit/importer.cpp]: The VM should not be returning `CORINFO_INTRINSIC_ReadOnlySpan_GetItem` for indexer that returns struct today. Sounds like something else is off...

**Comment by benaadams** [comment]
Related? "Regression in Span indexer codegen" https://github.com/dotnet/coreclr/issues/21032

**Comment by jkotas** [comment]
Yes, this looks like a regression introduced by the new Span indexers.
cc @tarekgh

**Comment by AndyAyersMS** [comment]
Do we (now) need a signature in mscorlib.h so we match the right overload?
```
DEFINE_METHOD(SPAN,                 GET_ITEM,               get_Item, NoSig)
DEFINE_METHOD(READONLY_SPAN,        GET_ITEM,               get_Item, NoSig)
```

**Comment by CarolEidt** [comment]
> Do we (now) need a signature in mscorlib.h so we match the right overload?

That would seem to be necessary if we want this to be handled in the VM. I'm not sure what the implications would be of simply failing to expand it in the JIT. AFAICT there aren't any current intrinsics that require signature matching.

**Comment by pentp** [comment]
Can the `[Intrinsic]` attribute be used to find the right indexer method and/or exclude the other indexers?

**Comment by CarolEidt** [comment]
> Can the [Intrinsic] attribute be used to find the right indexer method and/or exclude the other indexers?

As far as I know, the `[Intrinsic]` attribute is an actual attribute, i.e. not a special attribute encoded in the metadata, so it could be costly to check that, and probably easier to add sig support. (Though I'm still not sure that disqualifying the form that takes a struct isn't the cheapest and most expedient option.)

**Comment by jkotas** [comment]
>  can someone enlighten me about the dotnet.coreclr job and why it might have failed?

This is known issue - folks working on the new Azure DevOps infra. cc @morganbr

**Comment by morganbr** [comment]
Ignore the failure. Something got turned on that wasn't supposed to be.

**Comment by jkotas** [comment]
> Do we (now) need a signature in mscorlib.h so we match the right overload?

Yes, we need VM to match the right overload. The overload returned by the VM is non-deterministic right. It can either return the `int` overload or `Range` overload depending on multiple different factors.

**Comment by jkotas** [comment]
Added signatures to corelib: #21048

**Comment by tarekgh** [comment]
Is there anything need to be done from my side?

---
In reply to: [439243230](https://github.com/dotnet/coreclr/pull/21037#issuecomment-439243230) [](ancestors = 439243230)

**Comment by jkotas** [comment]
> Is there anything need to be done from my side?

- For this fix (#21048): Help to figure our whether and how this needs to get to preview1
- For the Indexer/Range: Make sure that the https://github.com/dotnet/corefx/issues/33376 follow up  does not fall through the cracks

**Comment by CarolEidt** [comment]
@AndyAyersMS - I've modified this PR to just add the test as well as a couple of asserts.
cc @dotnet/jit-contrib

**Pull request merged**